### ISSUE TITLE
Simulating CO2 air concentrations

### ISSUE THREAD

**Issue by Mathadon** [iea-annex60/modelica-annex60#372]
I intend to start simulating CO2 concentrations in my zone models. Should we look into making some standard components for this too? For instance an 'occupant' model, which emits heat (latent and co2) and co2 could be useful.

**Comment by mwetter** [comment]
That would be useful.

We currently have `Annex60.Fluid.Sources.TraceSubstancesFlowSource`. However, for adding CO2 or moisture to the volume, I would rather use a `RealInput`. The reason is that otherwise, you have for a mixing volume with inlet and outlet three or more fluid ports (inlet, outlet plus CO2 or moisture) and hence you can no longer use `allowFlowReversal=false` as the flow direction is not clear if a vectorized port is used.

In `Buildings.Rooms.MixedAir` we use `Buildings.Rooms.BaseClasses.MixedAirHeatGain` but the connector `QLat_flow` should really be a `RealOutput` and an optional output for `CO2` should be added.

**Comment by Mathadon** [comment]
I agree that a `RealInput` in a `MixingVolume` would be useful for the reasons you mention. Should we then make a component `MixingVolumeTraceSubstance` similar to `MixingVolumeMoistAir` that includes an input for moist air and for trace substances? It might be better to create a single `MixingVolume` model and just create conditional inputs?

**Comment by mwetter** [comment]
@Mathadon : Can you please review my changes on the branch `issue372_CO2`.
I did the following 
- corrected a bug in the mass fraction of which you used the inverse ratio. See also the regression tests `Fluid/Sensors/Examples/TraceSubstances.mo`
- refactored the volumes to avoid mixing assignments and connect statements in the mixing volume
- removed the unit of C, which in general need not be kg/s. (However, for the ppm sensor to be corrected, C must be in units of kg/s.
- Corrected the steady-state balance in the conservation equation model

All regression tests work.
We may want to consider renaming the sensor from `TraceSubstancesPpm` to `PPM`. In my view, it is clear that PPM is a trace substance.

**Comment by Mathadon** [comment]
Please check https://github.com/iea-annex60/modelica-annex60/pull/378/files

I implemented your first suggestion. I am not sure how to implement the second suggestion?

**Comment by mwetter** [comment]
@Mathadon : I added units in 94af57e

- [ ] If you agree with the implementation, then we should also add `PPMTwoPort` and add it to the unit tests. Best would be to make a new unit test so that users can easily find it.

**Comment by mwetter** [comment]
@Mathadon : Adding trace substances as an input should be done in `PartialMixingVolume` with conditionally enabled connectors as you suggested. This way, it will be available in `MixingVolume` and `MixingVolumeMoistAir`.  I don't like having `MixingVolume` and `MixingVolumeMoistAir`, but this was needed because `MixingVolumeMoistAir` requires `Medium.enthalpyOfLiquid` which is only available in moist air models, hence a separate model was needed to make sure that the enthalpy is consistent with the enthalpy used by the medium.

**Comment by Mathadon** [comment]
I'm currently a bit busy but I'll look into the two port later!

**Comment by Mathadon** [comment]
I think this is ready to be merged.

**Comment by mwetter** [comment]
@Mathadon 
I changed `use_C_flow_in` to `use_C_flow` as the connector is called `C_flow` and not `C_flow_in`.

Note that `Fluid.Sources` uses `C_flow_in`, but here I prefer to use `C_flow` as otherwise we would also need to change `Q_flow` to `Q_flow_in`. This apparent inconsistency with `Fluid.Sources` is in my view fine because `Fluid.Sources` uses also `C` as a parameter, which for these models does not make sense.

I also changed the connectors `C_flow` and `mWat_flow` to be conditional connectors as this is the modeling patterns that is typically used in the library. It also allowed to remove `masExc` which always outputted `0` and was now used for moisture and for trace substances which I think is confusing as one output is used for two different things, once for 0 kg/s, the other times to 0 1/s as trace substances need not be in units of kg. This may in fact be invalid as it connects `mWat_flow` to `C_flow` through `masExc.y` but they have different units.

- [ ] For some reason, you moved `tauInv` from the two port temperature sensor to its base class. I am fine with this change, but only if you change all other sensors to also use `tauInv`. Otherwise, we introduce a variable that is not used in the model that extend it (which is bad style). Chaning `xxx / tau` to `xxx * tauInv` may also be faster as no save division is required.

This is still needed (from my side):
- [ ] Further review models and the documentation.
- [ ] Review the changes in `Fluid.Sensors` from https://github.com/iea-annex60/modelica-annex60/pull/392/files.
- [ ] Run the unit tests.

**Comment by Mathadon** [comment]
I fixed a typo in `use_C_flow` in `Examples/PPM.mo`.
I used `tauInv` in the other sensors.

**Comment by mwetter** [comment]
I further revised the models to fix a previously broken unit test, in particular with commit 360a991. The issue was that the moisture was no longer added to the conservation equations in `Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation` because `m_flowInv=0`.
This required the following **backwards incompatible change:**

In `Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation`, the constant `sensibleOnly` was removed as the model uses now the parameter `use_mWat_flow` to enable/disable the input connector `mWat_flow`. Hence, `sensibleOnly` is no longer used by the model and has been removed.

The conversion script is:
```
convertModifiers("Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation", 
  {"sensibleOnly = true"}, fill("", 0), true);
convertModifiers("Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation", 
  {"sensibleOnly = false"}, {"use_mWat_flow = true"}, true);
```

**Comment by Mathadon** [comment]
I'd like to review this change, hopefully next week!

**Comment by mlauster** [comment]
Merged with #393
